### PR TITLE
upgrade development dependencies

### PR DESCRIPTION
**Pull request by MidnightWonderer**
In this PR I just use `npm audit` to resolve security vulnerabilities.  
  
I can resolve most of them with npm alone,
but by resolving gulp related vulnerable dependencies, gulp itself needs to be upgraded.  
Thus I also juggling around gulp config to make it compatible with gulp 4 requirements. (Composite tasks cannot be declared before the concrete implementation of tasks.)  
  
I am aware that this pull request will conflicts with others (including my own PR) just ping me whenever you feel like merging. I will rebase when needed.

### CODE REVIEW

**Comment by bojeil-google** [comment]
Hey @MidnightWonderer, thank you for your contributions. This PR doesn't build anymore, please resolve the open issues before we can help review this PR. Otherwise, we can try to build a fix based on your PR but that will probably take some time to get to.

**Comment by MidnightWonderer** [comment]
I'll look into it.  
I encountered the same issue once locally but it is transient and I am not able to reproduce it consistently.  
So I may trigger travis build from my branch multiple times to debug the issue.

**Comment by MidnightWonderer** [comment]
Now things get weirder.  
I cannot even reproduce it on travis anymore!  
  
All I did compare to the previous build was just npm upgrading.  
  
Do you want to try something else with the travis? (e.g. remove the last commit)  
Maybe it is related to timing but I cannot reproduce it locally. (albeit seen it once)

**Comment by wti806** [comment]
@MidnightWonderer We disabled the Saucelabs Test on travis for now. You can merge the change on master branch to your feature branch and run travis test again. And is the last commit just for testing?

**Comment by MidnightWonderer** [comment]
Yes just for testing. I removed the commit already.  
I am not able to reproduce the error even without the commit.  
  
For the sake of future reference, I will leave the link to the failing build here:  
https://travis-ci.org/firebase/firebaseui-web/builds/408566331

File: gulpfile.js [hunk 1]
```diff
@@ -164,18 +164,19 @@ function getTmpJsPath(locale) {
...
     const replaceTokens = (name) => name.replace(/\$/g, locale);
-
     const localeTaskName = replaceTokens(taskName);
-    tasks.push(localeTaskName);
     const localeDependencies = dependencies.map(replaceTokens);
 
-    gulp.task(localeTaskName, localeDependencies, () => operation(locale));
+    gulp.task(localeTaskName, gulp.series(
+      gulp.parallel(...localeDependencies),
+      () => operation(locale),
```
Comment (bojeil-google) [inline gulpfile.js]: Our style guide uses 4 spaces when listing parameters after a new line:
```
func(
    a,
    b);
```
Please update here and below.

File: gulpfile.js [hunk 2]
```diff
@@ -263,22 +252,44 @@ gulp.task('build-ts',
...
-// Builds the final JS file for the default language.
-gulp.task('build-js', ['build-js-' + DEFAULT_LOCALE],
-    () => makeDefaultFile('firebaseui'));
+// Builds the core FirebaseUI JS. Generates the gulp tasks
+// build-firebaseui-js-de, build-firebaseui-js-fr, etc.
+repeatTaskForAllLocales(
+  'build-firebaseui-js-$',
+  ['build-externs', 'build-ts', 'build-soy'],
+  buildFirebaseUiJs,
+);
```
Comment (bojeil-google) [inline gulpfile.js]: This looks unchanged. Is there a reason why you moved it? If not, let's revert this and keep the changes needed.

File: gulpfile.js [hunk 3]
```diff
@@ -263,22 +252,44 @@ gulp.task('build-ts',
...
 // the NPM module for all languages.
 repeatTaskForAllLocales(
     'build-npm-$', ['build-firebaseui-js-$'],
     (locale) => concatWithDeps(locale, 'npm', NPM_MODULE_WRAPPER));
 
+// Bundles the FirebaseUI JS with its dependencies for all locales.
+// Generates the gulp tasks build-js-de, build-js-fr, etc.
+const buildJsTasks = repeatTaskForAllLocales(
+  'build-js-$', ['build-firebaseui-js-$'],
+  (locale) => concatWithDeps(locale, 'firebaseui', OUTPUT_WRAPPER));
```
Comment (bojeil-google) [inline gulpfile.js]: This is also unchanged.

File: gulpfile.js [hunk 4]
```diff
@@ -322,17 +333,19 @@ gulp.task('serve', () => {
...
-      'build-css-rtl'
-    ],
-    () => gulp.start('clean'));
+gulp.task('default', gulp.series(
+  gulp.parallel(
+    'build-externs', 'build-ts', 'build-js',
+    'build-npm', 'build-css', 'build-css-rtl',
+  ),
+  'clean',
+));
```
Comment (bojeil-google) [inline gulpfile.js]: Apply 4 space indents.

File: gulpfile.js [hunk 5]
```diff
@@ -322,17 +333,19 @@ gulp.task('serve', () => {
...
-      'build-css-rtl'
-    ],
-    () => gulp.start('clean'));
+gulp.task('build-all', gulp.series(
+  gulp.parallel(
+    'build-externs', 'build-ts', 'build-all-js',
+    'build-npm', 'build-css', 'build-css-rtl',
+  ),
+  'clean',
+));
```
Comment (bojeil-google) [inline gulpfile.js]: Apply 4 space indents.

**Comment by MidnightWonderer** [comment]
OK, I just apply 4 space indentations to function invokation parameters.  
I'll explain the rest;  
why so many relocated code without actual functionality changes.  
  
In gulp 4 they are getting rid of cyclic task dependencies.  
By doing so they don't let you reference to named tasks without the task being declared.  
e.g.
You cannot put
```javascript
// Builds the core FirebaseUI JS. Generates the gulp tasks	
// build-firebaseui-js-de, build-firebaseui-js-fr, etc.	
repeatTaskForAllLocales('build-firebaseui-js-$',	
    ['build-externs', 'build-ts', 'build-soy'],	
    buildFirebaseUiJs);
```
before
```javascript
gulp.task('build-externs',
    () => gulp.src(EXTERNS_FILES).pipe(gulp.dest(`${DEST_DIR}/externs/`)));
```
because 'build-externs' was referenced before the declaration.  
  
I just randomly relocated the tasks based on this constraint.  
Please tell me if you have preferred order.  
However, the current one at the master branch won't do.

**Comment by MidnightWonderer** [comment]
I made a Travis build demo at:  
https://github.com/firebase/firebaseui-web/pull/472

**Comment by bojeil-google** [comment]
Interesting. Thanks for the explanation. I didn't expect the order would matter.

**Pull request merged**